vg/vgimg/vgimg.go

```diff
@@ -40,46 +44,73 @@ type Canvas struct {
 // New returns a new image canvas with
 // the size specified  rounded up to the
 // nearest pixel.
-func New(width, height vg.Length) *Canvas {
-	w := width / vg.Inch * dpi
-	h := height / vg.Inch * dpi
+func New(width, height vg.Length, options ...func(*Canvas)) *Canvas {
+	c := new(Canvas)
+	c.dpi = defaultDPI
+	for _, option := range options {
+		option(c)
+	}
+	w := width / vg.Inch * vg.Length(c.dpi)
+	h := height / vg.Inch * vg.Length(c.dpi)
 	img := image.NewRGBA(image.Rect(0, 0, int(w+0.5), int(h+0.5)))
 
-	return NewImage(img)
+	return NewImage(img, options...)
 }
 
 // NewImage returns a new image canvas
 // that draws to the given image.  The
 // minimum point of the given image
 // should probably be 0,0.
-func NewImage(img draw.Image) *Canvas {
+func NewImage(img draw.Image, options ...func(*Canvas)) *Canvas {
 	h := float64(img.Bounds().Max.Y - img.Bounds().Min.Y)
+	c := new(Canvas)
+	c.dpi = defaultDPI
+	for _, option := range options {
+		option(c)
+	}
 	gc := draw2d.NewGraphicContext(img)
-	gc.SetDPI(dpi)
+	gc.SetDPI(c.dpi)
 	gc.Scale(1, -1)
 	gc.Translate(0, -h)
-	return NewImageWithContext(img, gc)
+	return NewImageWithContext(img, gc, options...)
 }
 
 // NewImageWithContext returns a new image canvas
 // that draws to the given image, using the given graphic context.
 // The minimum point of the given image
 // should probably be 0,0.
-func NewImageWithContext(img draw.Image, gc draw2d.GraphicContext) *Canvas {
+func NewImageWithContext(img draw.Image, gc draw2d.GraphicContext,
+	options ...func(*Canvas)) *Canvas {
 	w := float64(img.Bounds().Max.X - img.Bounds().Min.X)
 	h := float64(img.Bounds().Max.Y - img.Bounds().Min.Y)
 	draw.Draw(img, img.Bounds(), image.White, image.ZP, draw.Src)
-	c := &Canvas{
-		gc:    gc,
-		img:   img,
-		w:     vg.Length(w/dpi) * vg.Inch,
-		h:     vg.Length(h/dpi) * vg.Inch,
-		color: []color.Color{color.Black},
+	c := new(Canvas)
+	c.dpi = defaultDPI
+	for _, option := range options {
+		option(c)
 	}
+	c.gc = gc
+	c.img = img
+	c.w = vg.Length(w/float64(c.dpi)) * vg.Inch
+	c.h = vg.Length(h/float64(c.dpi)) * vg.Inch
+	c.color = []color.Color{color.Black}
 	vg.Initialize(c)
+	c.initialized = true
 	return c
 }
 
+// DPI sets the dots per inch of a canvas. It should only be
+// used as an "option" argument when initializing a new canvas.
+func DPI(dpi int) func(*Canvas) {
+	return func(c *Canvas) {
+		if c.initialized {
```

Comment:
I think that you are trying to avoid someone doing:
```c := vgimg.New(...)
DPI(12345)(c)```
which is invalid because it changes the DPI without changing it on the underlying GraphicsContext. A different way to prevent this is to make options opaque:
```type option func(*Canvas)
func DPI(dpi int) option { ... }```
The caller can still create an option via `DPI()` and pass it to `New`, but now they don't know what it is and they cannot call it themselves, and you don't need the initialized flag or this panic.